File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    ArrayList<Integer>[] graph;
    boolean[] visited;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        graph = new ArrayList[n];
        visited = new boolean[n];
        
        for(int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(computers[i][j] == 1 && i != j) {
                    graph[i].add(j);
                }
            }
        }
        
        for(int i = 0; i < n; i++) {
            if(!visited[i]) {
                answer++;
                dfs(i);
            }
        }
        
        return answer;
    }
    
    void dfs(int cur) {
        visited[cur] = true;
        for(Integer node: graph[cur]) {
            if(!visited[node]) {
                visited[node] = true;
                dfs(node);
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def bfs(graph, visited, v):
    queue = deque()
    queue.append(v)
    
    while queue:
        node = queue.popleft()
        for next in graph[node]:
            if not visited[next]:
                queue.append(next)
                visited[next] = True
    
    return visited
    

def solution(n, computers):
    answer = 0
    
    graph = [[] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] == 1:
                graph[i].append(j)
    
    visited = [False] * n
    for i in range(n):
        if not visited[i]:
            visited = bfs(graph, visited, i)
            answer += 1
    
    return answer